### ISSUE TITLE
training procedure and mAP

### ISSUE THREAD

**Issue by GotG** [zylo117/Yet-Another-EfficientDet-Pytorch#232]
700 images dataset, 3 classes, trained for 150 epocshs - EffDet D0
train total loss: 0.37. val total loss 0.34
inference looks ok. 
however the coco evaluation results are quite bad:

Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.076
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.140
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.082
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.000
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.045
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.153
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.030
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.094
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.125
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.000
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.090
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.227

what is wrong?

**Comment by zylo117** [comment]
care to share your training commands?

**Comment by zhicai-liu** [comment]
I encountered the same problem

**Comment by GotG** [comment]
> 
> 
> care to share your training commands?

sure! 
! python train.py -c 0 -p hardhat --head_only True --lr 1e-3 --batch_size 32 --load_weights weights/efficientdet-d0.pth  --num_epochs 50
then i change lr to 1e-4 and 1e-6 until epoch 150. after 150 i get no improvement.

Step: 3685. Epoch: 193/200. Iteration: 19/19. Cls loss: 0.34229. Reg loss: 0.02093. Total loss: 0.36323: 100% 19/19 [01:01<00:00,  3.22s/it]
Val. Epoch: 193/200. Classification loss: 0.30842. Regression loss: 0.02742. Total loss: 0.33584

**Comment by zylo117** [comment]
It's not overfitting.And the loss is low enough.
Is it the eval result always the same?
Is it the boxes' shape far different from coco's, i.e. long rectangles?

**Comment by GotG** [comment]
yea, the losses look ok. yes eval result always the same. i tried deleting the benchmark file, still same result. so i think some bug there.
also confidence threshold for inference is quite low - ard 0.4 for the above model.

**Comment by zylo117** [comment]
did you load the same weights every time?

if mAP is only 0.07, are there lots of valid boxes?

did you use float16? don't.

**Comment by kuangsangudu** [comment]
My problem is that my reg loss is very small from the beginning, and it has basically not changed. The cls loss is gradually decreasing, but the regression effect is still not very good when I check the model. 
My training set is more than 700 pictures (the training set I modified slightly and it is the traffic sign data set), and I trained for 100 epochs.

**Comment by zylo117** [comment]
I don't think you can train on 700 images and get a good result. It's not enough. Especially for such a  deep network like effdet, the training will be very unstable and overfitting.

**Comment by GotG** [comment]
i think the problem is the class imbalance. i have 700 images and the categories split like this:
614 , 2859, 120
the ssd models do a decent job, getting all the way to a 67 mAP on a summarily trained retinanet.
the EffDet network overfits heavily the class. 
so it would benefit from a hard miner/focal loss presence.
so i should first look to have the classes balanced. any other suggestions?
maybe i can change the loss to focal loss? thanks for repo and help!

**Comment by zylo117** [comment]
it is focal loss already.
and yes, effdet is quite a deep network and is hard to train, much harder than ssd and yolo.
According to your loss, I say it's not even converged, far from it. Just keep training.
BTW, my final val cls loss for coco on D0 is 0.35. Considering there are 90 classes in coco, it should be much lower on a 3-classes dataset.

**Comment by GotG** [comment]
because of this class imbalance it severly overfits. maybe it corrects if i allow it to train more. i will go for 1e-3 for 500 epochs bs =32

**Comment by kuangsangudu** [comment]
i  want to know how long it takes to train D0 by 2080.

**Comment by GotG** [comment]
Training update

! python train.py -c 0 -p medmask --head_only True --lr 2e-3 --batch_size 32 --load_weights weights/efficientdet-d0.pth  --num_epochs 500

No change btw epoch 177 and 220 (where it is now)

Step: 3381. Epoch: 177/500. Iteration: 19/19. Cls loss: 0.33934. Reg loss: 0.01636. Total loss: 0.35571: 100% 19/19 [00:49<00:00,  2.61s/it]
Val. Epoch: 177/500. Classification loss: 0.31454. Regression loss: 0.02704. Total loss: 0.34158
Step: 4198. Epoch: 220/500. Iteration: 19/19. Cls loss: 0.36939. Reg loss: 0.01953. Total loss: 0.38892: 

100% 19/19 [00:50<00:00,  2.65s/it]
Val. Epoch: 220/500. Classification loss: 0.31435. Regression loss: 0.02707. Total loss: 0.34143

**Comment by zylo117** [comment]
hold on, did you load the same coco-pretrained weights?
that's why you are not improving. you should load the last weights. `--load_weights last`

**Comment by zylo117** [comment]
and val reg loss is much higher than train reg loss. you should visualize them on tensorboard. it helps.

**Comment by GotG** [comment]
my procedure is simple: 
load the effdet d0 pretrained and let it run for 500 epochs.
what i pasted above is from ongoing training. 
as discussed above you suggested to let it train for a long time. so i started from scratch and going for 500 epochs.
i know it is overfitting on one class since i have the class imbalance in my training data.
also on prior runs on inference it was detecting only one class, ntg else. so that's the visual proof as well.
focal loss is implemented like you said but unfortunately it does not seem to help.
looks like this model can not handle class imbalance for me.

**Comment by Cli98** [comment]
> my procedure is simple:
> load the effdet d0 pretrained and let it run for 500 epochs.
> what i pasted above is from ongoing training.
> as discussed above you suggested to let it train for a long time. so i started from scratch and going for 500 epochs.
> i know it is overfitting on one class since i have the class imbalance in my training data.
> also on prior runs on inference it was detecting only one class, ntg else. so that's the visual proof as well.
> focal loss is implemented like you said but unfortunately it does not seem to help.
> looks like this model can not handle class imbalance for me.

@GotG Did you modify your anchor? In case you want to try, check this repo https://github.com/Cli98/anchor_computation_tool
